### PR TITLE
Caching implmented using reedis

### PR DESCRIPTION
**Pull request by obedjamir**
getStatusByName api implemented with @Cacheable
Cache time to live set to 5 seconds
test cases iplemented for cacheable

### CODE REVIEW

File: core/src/main/java/com/bankbazaar/kafka/core/model/FileStatusEntity.java [hunk 1]
```diff
@@ -16,7 +16,7 @@ public class FileStatusEntity {
     @GeneratedValue
     private Long id;
 
-    @Column(name="file_name",nullable = false)
+    @Column(name="fileName",nullable = false)
```
Comment (PinakiGhosh) [inline core/src/main/java/com/bankbazaar/kafka/core/model/FileStatusEntity.java]: Why is this change required?

File: core/src/main/java/com/bankbazaar/kafka/core/model/Status.java [hunk 2]
```diff
@@ -1,5 +1,5 @@
 package com.bankbazaar.kafka.core.model;
 
 public enum Status {
-    NEW,IN_PROGRESS,SUCCESS,FAILURE,ERROR
+    NEW,IN_PROGRESS,SUCCESS,FAILURE,ERROR,NULL
```
Comment (PinakiGhosh) [inline core/src/main/java/com/bankbazaar/kafka/core/model/Status.java]: NULL should not be a status

File: core/src/main/java/com/bankbazaar/kafka/core/manager/FileStatusManager.java [hunk 3]
```diff
@@ -34,4 +35,13 @@ public Optional<FileStatusEntity> getEntry(Long id)
     {
         return fileStatusRepository.findById(id);
     }
+    public Status getEntry(String name)
+    {
+        FileStatusEntity response = fileStatusRepository.findByFileName(name);
+        if(response==null)
+        {
+            return Status.NULL;
```
Comment (PinakiGhosh) [inline core/src/main/java/com/bankbazaar/kafka/core/manager/FileStatusManager.java]: Just return null

File: core/src/main/java/com/bankbazaar/kafka/core/manager/FileStatusManager.java [hunk 4]
```diff
@@ -34,4 +35,13 @@ public Optional<FileStatusEntity> getEntry(Long id)
     {
         return fileStatusRepository.findById(id);
     }
+    public Status getEntry(String name)
+    {
+        FileStatusEntity response = fileStatusRepository.findByFileName(name);
```
Comment (PinakiGhosh) [inline core/src/main/java/com/bankbazaar/kafka/core/manager/FileStatusManager.java]: Make the repository return an Optional

File: service/src/main/java/com/bankbazaar/kafka/service/config/CacheConfig.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+@Configuration
+@EnableCaching
+public class CacheConfig {
+
+    @Bean
+    public RedisCacheManagerBuilderCustomizer redisCacheManagerBuilderCustomizer() {
+        return (builder) -> builder
+                .withCacheConfiguration("StatusCache",
+                        RedisCacheConfiguration.defaultCacheConfig().entryTtl(Duration.ofSeconds(5)));
```
Comment (PinakiGhosh) [inline service/src/main/java/com/bankbazaar/kafka/service/config/CacheConfig.java]: Add the ttl as a config. This way you can have separate values for actual application and test

File: service/src/main/java/com/bankbazaar/kafka/service/service/RedisUtil.java [hunk 6]
```diff
@@ -14,7 +14,7 @@ public class RedisUtil {
 
     public void saveToRedis(Long id, Status status)
     {
-        redisTemplate.opsForValue().set(id, status, 10, TimeUnit.SECONDS);
+        redisTemplate.opsForValue().set(id, status, 5, TimeUnit.SECONDS);
```
Comment (PinakiGhosh) [inline service/src/main/java/com/bankbazaar/kafka/service/service/RedisUtil.java]: Why is this required?

File: service/src/test/java/com/bankbazaar/kafka/service/controller/KafkaControllerTest.java [hunk 7]
```diff
@@ -90,10 +103,16 @@ void userController() throws Exception {
         DataDto dataDto4 = createBadFileObject();
         consumeBadApi(dataDto4);
 
-        Thread.sleep(10000);
+        Thread.sleep(5000);
```
Comment (PinakiGhosh) [inline service/src/test/java/com/bankbazaar/kafka/service/controller/KafkaControllerTest.java]: You should not be using sleep anywhere

**Review by PinakiGhosh — CHANGES_REQUESTED** [review]

File: core/src/main/java/com/bankbazaar/kafka/core/model/FileStatusEntity.java (hunk 1 above)
Comment (obedjamir) [reply core/src/main/java/com/bankbazaar/kafka/core/model/FileStatusEntity.java]: I have reverted the change

File: core/src/main/java/com/bankbazaar/kafka/core/model/Status.java (hunk 2 above)
Comment (obedjamir) [reply core/src/main/java/com/bankbazaar/kafka/core/model/Status.java]: NULL removed

File: core/src/main/java/com/bankbazaar/kafka/core/manager/FileStatusManager.java (hunk 3 above)
Comment (obedjamir) [reply core/src/main/java/com/bankbazaar/kafka/core/manager/FileStatusManager.java]: method now returns null

File: core/src/main/java/com/bankbazaar/kafka/core/manager/FileStatusManager.java (hunk 4 above)
Comment (obedjamir) [reply core/src/main/java/com/bankbazaar/kafka/core/manager/FileStatusManager.java]: repository returns optional

File: service/src/main/java/com/bankbazaar/kafka/service/config/CacheConfig.java (hunk 5 above)
Comment (obedjamir) [reply service/src/main/java/com/bankbazaar/kafka/service/config/CacheConfig.java]: Ttl added as config

File: service/src/main/java/com/bankbazaar/kafka/service/service/RedisUtil.java (hunk 6 above)
Comment (obedjamir) [reply service/src/main/java/com/bankbazaar/kafka/service/service/RedisUtil.java]: This was done to minimize time during testing. However, ttl can now be configured seperately

File: service/src/test/java/com/bankbazaar/kafka/service/controller/KafkaControllerTest.java (hunk 7 above)
Comment (obedjamir) [reply service/src/test/java/com/bankbazaar/kafka/service/controller/KafkaControllerTest.java]: this has been removed

File: service/src/test/java/com/bankbazaar/kafka/service/controller/KafkaControllerTest.java [hunk 8]
```diff
@@ -46,7 +45,10 @@
     void userController() throws Exception {
         ClassLoader classLoader = getClass().getClassLoader();
 
-        assertEquals(consumeApi("test2.csv"), Status.NULL);
+        /**
+         * controller returns null before any db entry
+         */
+        assertNull(consumeApi("test2.csv"));
```
Comment (PinakiGhosh) [inline service/src/test/java/com/bankbazaar/kafka/service/controller/KafkaControllerTest.java]: This is a scenario where the given filename does not exist right? Should it not then respond with a 404 status?

File: service/src/main/resources/application.properties [hunk 9]
```diff
@@ -30,4 +30,4 @@ spring.cloud.stream.bindings.consumer-out-0.destination=Notification
 spring.cloud.stream.bindings.notification-in-0.destination=Notification
 spring.datasource.maxRetries=1
 spring.cache.type=redis
-
+spring.datasource.timeToLive=259200
```
Comment (PinakiGhosh) [inline service/src/main/resources/application.properties]: Mention the unit as well in the properties

fileStatus.cache.expiry.inseconds=

Avoid using spring.* since those have are generally for the spring boot framework

**Review by PinakiGhosh — CHANGES_REQUESTED** [review]

**Review by PinakiGhosh — APPROVED** [review]
